### PR TITLE
lua: expose full interface of vim.inspect and add test

### PR DESCRIPTION
**Pull request by bfredl**
The current approach of lazy-loading `inspect` breaks the interface, fix it and add a test.

Implement lazy loading for vim.submodule, this would be over-engineering for `inspect` only, but we expect to use this solution also for more and larger modules.

### CODE REVIEW

File: src/nvim/lua/vim.lua [hunk 1]
```diff
@@ -195,6 +187,13 @@ deepcopy = function(orig)
   return deepcopy_funcs[type(orig)](orig)
 end
 
+local function __index(table, key)
+  if key == "inspect" then
+    table.inspect = require("vim.inspect")
+    return table.inspect
```
Comment (justinmk) [inline src/nvim/lua/vim.lua]: does `require()` do its own caching? If not, shouldn't we avoid it after `table.inspect` is non-nil?

Comment (bfredl) [reply src/nvim/lua/vim.lua]: @justinmk the line above does the caching.

Comment (justinmk) [reply src/nvim/lua/vim.lua]: Thanks. For [reference](https://stackoverflow.com/a/18026617/):

> `__index` is only accessed when the key being read from the main table does not already exist in the main table

When the key being read from the main table does not already exist in the main table. 

Also [this](https://stackoverflow.com/a/34235619/) explains `require()`:

> `require` caches loaded modules. Calling it again with the same module name will simply return the module that was previously loaded

**Review by justinmk — APPROVED** [review]

File: src/nvim/lua/vim.lua (hunk 1 above)
Comment (bfredl) [reply src/nvim/lua/vim.lua]: section 2.8 in https://www.lua.org/manual/5.1/manual.html contains functionalistic descriptions for how the metatable mechanism works.

**Pull request merged**